xls/tools/eval_ir_main.cc

```diff
@@ -528,20 +528,16 @@ absl::Status Run(Package* package, absl::Span<const ArgSet> arg_sets_in) {
       absl::GetFlag(FLAGS_output_node_coverage_stats_proto).has_value() ||
       absl::GetFlag(FLAGS_output_node_coverage_stats_textproto).has_value();
 
-  // TODO: google/xls#3307 - remove when multiple observers are supported.
-  if (trace_enabled && coverage_enabled) {
-    return absl::InvalidArgumentError(
-        "Cannot specify both --trace_output and node coverage flags.");
-  }
-
   ScopedRecordNodeCoverage cov(
       coverage_enabled ? absl::GetFlag(FLAGS_output_node_coverage_stats_proto)
                        : std::nullopt,
       coverage_enabled
           ? absl::GetFlag(FLAGS_output_node_coverage_stats_textproto)
           : std::nullopt);
   std::unique_ptr<ScopedTracingObserver> tracing_observer;
+  std::unique_ptr<CompositeEvaluationObserver> composite_observer;
```

Comment:
Make `CompositeEvaluationObserver` modifable as mentioned above then you can just leave this as a stack object and add stuff to it.